test/functional/fixtures/regression/gh-4486/testcafe-fixtures/index.js

```diff
@@ -0,0 +1,66 @@
+import { ClientFunction, Selector } from 'testcafe';
+
+fixture `GH-4486 - The 'beforeinput' event should be raised`
+    .page `http://localhost:3000/fixtures/regression/gh-4486/pages/index.html`;
+
+const getEventLog = ClientFunction(() => window.events);
+
+const expectedAllEvents = [
+    { type: 'beforeinput', data: '1' },
+    { type: 'textInput', data: '1' },
+    { type: 'beforeinput', data: '2' },
+    { type: 'textInput', data: '2' },
+    { type: 'beforeinput', data: '3' },
+    { type: 'textInput', data: '3' },
+];
+
+const expectedOnlyBeforeInput = [
+    { type: 'beforeinput', data: '1' },
+    { type: 'beforeinput', data: '2' },
+    { type: 'beforeinput', data: '3' }
+];
+
+const expectedOnlyTextInput = [
+    { type: 'textInput', data: '1' },
+    { type: 'textInput', data: '2' },
+    { type: 'textInput', data: '3' }
+];
+
+test(`beforeinput`, async t => {
+    await t.typeText('#input1', '123');
+
+    const log = await getEventLog();
+
+    if (t.browser.alias === 'chrome')
+        await t.expect(log).eql(expectedAllEvents);
+    if (t.browser.alias === 'safari')
+        await t.expect(log).eql(expectedAllEvents);
+
+    await t.expect(Selector('#input1').value).eql('123');
+});
+
+test(`Prevent beforeinput`, async t => {
+    await t.typeText('#input2', '123');
+
+    const log = await getEventLog();
+
+    if (t.browser.alias === 'chrome')
+        await t.expect(log).eql(expectedOnlyBeforeInput);
+    if (t.browser.alias === 'safari')
+        await t.expect(log).eql(expectedAllEvents);
+
+    await t.expect(Selector('#input2').value).eql('');
+});
+
+test(`Prevent textInput`, async t => {
```

Comment:
I think it would be better to write unit tests for the click automation instead of these functional ones

Reply:
It definitely would